src/main/java/com/redhat/devtools/lsp4ij/features/formatting/LSPEditorImprovementsTypedHandler.java

```diff
@@ -0,0 +1,208 @@
+            Matcher stringLiteralMatcher = stringLiteralPattern.matcher(fileChars);
+            while (stringLiteralMatcher.find()) {
+                int stringLiteralStart = stringLiteralMatcher.start();
+                int stringLiteralEnd = stringLiteralMatcher.end();
+                if ((stringLiteralStart < offset) && (offset < stringLiteralEnd)) {
+                    return fileChars.subSequence(stringLiteralStart, stringLiteralEnd).toString();
+                }
+            }
+        }
+
+        return null;
+    }
+
+    private static boolean inStringLiteral(@NotNull PsiFile file,
+                                           @NotNull CharSequence documentChars,
+                                           int offset) {
+        return getStringLiteral(file, documentChars, offset) != null;
+    }
+
+    private static boolean inComment(@NotNull PsiFile file,
+                                     @NotNull CharSequence documentChars,
+                                     int offset) {
+        boolean inComment = false;
+
+        Document document = LSPIJUtils.getDocument(file);
+        if (document != null) {
+            Commenter commenter = LSPIJEditorUtils.getCommenter(file);
+
+            // First try to see if we're in a line comment since that's much more efficient. We're in a lne
+            // comment if we find a line comment start on the same line before the current position.
+            String lineCommentPrefix = commenter.getLineCommentPrefix();
+            if (StringUtil.isNotEmpty(lineCommentPrefix)) {
+                int lineNumber = document.getLineNumber(offset);
+                int lineStartOffset = document.getLineStartOffset(lineNumber);
+                int lineEndOffset = document.getLineEndOffset(lineNumber);
+                if (lineEndOffset > lineStartOffset) {
+                    String lineText = documentChars.subSequence(lineStartOffset, lineEndOffset).toString();
+                    if (StringUtil.isNotEmpty(lineText)) {
+                        int lineCommentPrefixIndex = lineText.lastIndexOf(lineCommentPrefix, offset);
+                        if (lineCommentPrefixIndex > -1) {
+                            int lineCommentPrefixOffset = lineStartOffset + lineCommentPrefixIndex;
+                            if (lineCommentPrefixOffset < offset) {
+                                inComment = true;
+                            }
+                        }
+                    }
+                }
+            }
+
+            // If not detected in a line comment, try to see if we're in a block comment. We're in a block
+            // comment if we find a block comment start before the current position without a block comment
+            // end in the interim.
+            if (!inComment) {
+                String blockCommentPrefix = commenter.getBlockCommentPrefix();
```

Comment:
Block comment is pretty simple as well. Just see if there's a block comment prefix before the caret with no block comment suffix between the prefix and the caret. If so, the caret is in a block comment.